interfaces/prompting/prompting.go

```diff
@@ -0,0 +1,139 @@
+// the given expiration is valid for that lifespan.
+//
+// If the lifespan is LifespanTimespan LifespanTimespan, then expiration must
+// be non-nil. Otherwise, it must be nil. Returns an error if any of the above
+// are invalid.
+func ValidateLifespanExpiration(lifespan LifespanType, expiration *time.Time, currTime time.Time) error {
+	switch lifespan {
+	case LifespanForever, LifespanSession, LifespanSingle:
+		if expiration != nil {
+			return fmt.Errorf(`invalid expiration: expiration must be nil when lifespan is %q, but received non-nil expiration: %q`, lifespan, *expiration)
+		}
+	case LifespanTimespan:
+		if expiration == nil {
+			return fmt.Errorf(`invalid expiration: expiration must be non-nil when lifespan is %q, but received nil expiration`, lifespan)
+		}
+		if currTime.After(*expiration) {
+			return fmt.Errorf("invalid expiration: expiration time has already passed: %q", *expiration)
+		}
+	default:
+		return fmt.Errorf(`invalid lifespan: %q`, lifespan)
+	}
+	return nil
+}
+
+// ValidateLifespanParseDuration checks that the given lifespan is valid and
+// that the given duration is valid for that lifespan.
+//
+// If the lifespan is LifespanTimespan, then duration must be a string parsable
+// by time.ParseDuration(), representing the duration of time for which the rule
+// should be valid. Otherwise, it must be empty. Returns an error if any of the
+// above are invalid, otherwise computes the expiration time of the rule based
+// on the current time and the given duration and returns it.
+func ValidateLifespanParseDuration(lifespan LifespanType, duration string) (*time.Time, error) {
+	var expiration *time.Time
+	switch lifespan {
+	case LifespanForever, LifespanSession, LifespanSingle:
+		if duration != "" {
+			return nil, fmt.Errorf(`invalid duration: duration must be empty when lifespan is %q, but received non-empty duration: %q`, lifespan, duration)
+		}
+	case LifespanTimespan:
+		if duration == "" {
+			return nil, fmt.Errorf(`invalid duration: duration must be non-empty when lifespan is %q, but received empty expiration`, lifespan)
+		}
+		parsedDuration, err := time.ParseDuration(duration)
+		if err != nil {
+			return nil, fmt.Errorf(`invalid duration: error parsing duration string: %q`, duration)
+		}
+		if parsedDuration <= 0 {
+			return nil, fmt.Errorf(`invalid duration: duration must be greater than zero: %q`, duration)
+		}
+		expirationValue := time.Now().Add(parsedDuration)
+		expiration = &expirationValue
+	default:
+		return nil, fmt.Errorf(`invalid lifespan: %q`, lifespan)
+	}
+	return expiration, nil
+}
+
+// NewIDAndTimestamp returns a new unique ID and corresponding timestamp.
+//
+// The ID is the current unix time in nanoseconds encoded as a string in base32.
+// The timestamp is the same time, encoded as a string in time.RFC3339Nano.
+func NewIDAndTimestamp() (id string, timestamp time.Time) {
+	now := time.Now()
+	nowUnix := uint64(now.UnixNano())
```

Comment:
According to the time package documentation (https://pkg.go.dev/time#hdr-Monotonic_Clocks), `time.Now()` automatically includes both monotonic and wall time:

> If Times t and u both contain monotonic clock readings, the operations t.After(u), t.Before(u), t.Equal(u), t.Compare(u), and t.Sub(u) are carried out using the monotonic clock readings alone, ignoring the wall clock readings. If either t or u contains no monotonic clock reading, these operations fall back to using the wall clock readings.

> Because the monotonic clock reading has no meaning outside the current process, the serialized forms generated by t.GobEncode, t.MarshalBinary, t.MarshalJSON, and t.MarshalText omit the monotonic clock reading, and t.Format provides no format for it. Similarly, the constructors time.Date, time.Parse, time.ParseInLocation, and time.Unix, as well as the unmarshalers t.GobDecode, t.UnmarshalBinary. t.UnmarshalJSON, and t.UnmarshalText always create times with no monotonic clock reading.

Thus, the monotonic time is only stripped when rules/prompts are read from disk. This occurs when snapd is restarted, including on system start. Unfortunately, this means `time.Now()` is not guaranteed to be unique between snapd restarts, so we likely need to include both boot ID and system monotonic time when creating rule/prompt IDs.

Reply:
where and what are this IDs used for ? in general snapd prefers to use sequential increasing IDs? at least for norification from the kernel don't we get them in a single goroutine? in that context using sequential IDs would be cheap, no?

Reply:
AFAIU the IDs are stored within the prompting namespace in notices and we need them to be unique across reboots too (@olivercalder please correct me if I'm wrong), so to keep them sequential we'd either need to store some state.

Reply:
yes, that's a good point

Reply:
We could start from the max of the pending ones, though that would allow for the potential that IDs are reused for different rules, if a rule is removed, then snapd restarted. This isn't necessarily a problem, as we don't guarantee that IDs are never reused, but I'd prefer to keep them unique if possible.

I think storing a count as part of the rule storage would solve this problem, and it can be atomically incremented whenever a new rule is added. For prompt requests, which are not written to disk (because of listener/notify socket behavior, but that can be discussed in a later PR), we can just keep a count in memory and atomically increment it.

So, my opinion is that we should remove this function for now and add a `NewID` function when necessary in the requestrules and requestprompts packages.